pymodbus/transport/base.py

```diff
@@ -0,0 +1,109 @@
+"""Base for all transport types."""
+from __future__ import annotations
+
+from typing import Any, Callable
+
+from pymodbus.framer import ModbusFramer
+from pymodbus.logging import Log
+
+
+class BaseTransport:
+    """Base class for transport types.
+
+    :param framer: framer used to encode/decode data
+    :param slaves: list of acceptable slaves (0 for accept all)
+    :param comm_name: name of this transport connection
+    :param reconnect_delay: delay in milliseconds for first reconnect (0 for no reconnect)
+    :param reconnect_delay_max: max delay in milliseconds for next reconnect, resets to reconnect_delay
+    :param retries_send: number of times to retry a send operation
+    :param retry_on_empty: retry read on nothing
+    :param timeout_connect: Max. time in milliseconds for connect to complete
+    :param timeout_comm: Max. time in milliseconds for recv/send to complete
+    :param on_connection_made: callback
+    :param on_connection_lost: callback
+
+    :property reconnect_delay_current: current delay in milliseconds for next reconnect (doubles with every try)
+    :property transport: current transport class (none if not connected)
+
+    BaseTransport contains functions common to all transport types and client/server.
+
+    This class is not available in the pymodbus API, and should not be referenced in Applications.
+    """
+
+    def __init__(self) -> None:
+        """Initialize a transport instance."""
+        # parameter variables, overwritten in child classes
+        self.framer: ModbusFramer | None = None
+        self.slaves: list[int] = []
+        self.comm_name: str = ""
+        self.reconnect_delay: int = -1
+        self.reconnect_delay_max: int = -1
+        self.retries_send: int = -1
+        self.retry_on_empty: int = -1
+        self.timeout_connect: bool = None
+        self.timeout_comm: int = -1
+        self.on_connection_made: Callable[[str], None] = None
+        self.on_connection_lost: Callable[[str, Exception], None] | None = None
+
+        # local variables
+        self.reconnect_delay_current: int = 0
+        self.transport: Any = None
+
+    # -------------------------- #
+    # Transport external methods #
+    # -------------------------- #
+    def connection_made(self, transport: Any):
+        """Call from asyncio, when a connection is made.
+
+        :param transport: socket etc. representing the connection.
+        """
+        self.transport = transport
+        Log.debug("Connected {}", self.comm_name)
+        if self.on_connection_made:
+            self.on_connection_made(self.comm_name)  # pylint: disable=not-callable
```

Comment:
@alexrudd2 please do me a favor and have a look at this puzzle.

I declare on_connection_made in line 45, and pylint complains here in line 63 (same goes for on_connection_lost). I simply cannot make it right for both pylint and mypy, while keeping it working.

The idea is that the inherited classes can set a callback and get called, if the callback is None no action is taken.

I am sure it is something simple, but I have run out of ideas.

Thanks in advance for your help.

Reply:
Well, I don't know if it's a _good_ solution, but it works:

`self.on_connection_made: Callable[[str], None] = lambda foo : None`

Then delete the conditional and always call the callback.  😆

Reply:
Thanks, at least a solution.